store/s3storeKMS.go

```diff
@@ -0,0 +1,486 @@
+package store
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/pkg/errors"
+)
+
+// latest is used to keep a single object in s3 with all of the
+// most recent values for the given service's secrets.  Keeping this
+// in a single s3 object allows us to use a single s3 GetObject
+// for ListRaw (and thus chamber exec).
+type LatestIndexFile struct {
+	Latest map[string]LatestValue `json:"latest"`
+}
+
+type LatestValue struct {
+	Version  int    `json:"version"`
+	Value    string `json:"value"`
+	KMSAlias string `json:"KMSAlias"`
+}
+
+var _ Store = &S3KMSStore{}
+
+type S3KMSStore struct {
+	svc    s3iface.S3API
+	stsSvc *sts.STS
+	bucket string
+}
+
+// Deprecated; use NewS3KMSStoreWithBucket instead
+func NewS3KMSStore(numRetries int) (*S3KMSStore, error) {
+	bucket, ok := os.LookupEnv(BucketEnvVar)
+	if !ok {
+		return nil, fmt.Errorf("Must set %s for s3 backend", BucketEnvVar)
+	}
+
+	return NewS3KMSStoreWithBucket(numRetries, bucket)
+}
+
+func (s *S3KMSStore) KMSKey() string {
+	fromEnv, ok := os.LookupEnv("CHAMBER_KMS_KEY_ALIAS")
+	if !ok {
+		return DefaultKeyID
+	}
+	if !strings.HasPrefix(fromEnv, "alias/") {
+		return fmt.Sprintf("alias/%s", fromEnv)
+	}
+
+	return fromEnv
+}
+
+func NewS3KMSStoreWithBucket(numRetries int, bucket string) (*S3KMSStore, error) {
```

Comment:
The S3 store's long name is to maintain compatibility. You're starting fresh here.

```
type S3KMSStore struct {
	svc    s3iface.S3API
	stsSvc *sts.STS
	bucket string
	kmsKeyAlias string
}

func NewS3KMSStore(numRetries int, bucket string, kmsKeyAlias string) (*S3KMSStore, error) {
	if kmsKeyAlias == "" {
		kmsKeyAlias = DefaultKeyID
	}
	...
```
Then we can get rid of `KMSKey()` as well.

Reply:
👍